File: java/tengi-core/src/main/java/com/noctarius/tengi/spi/buffer/MemoryBuffer.java
/*
 * Copyright (c) 2015-2016, Christoph Engelbert (aka noctarius) and
 * contributors. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.noctarius.tengi.spi.buffer;

/**
 * The <tt>MemoryBuffer</tt> interface describes the combination of
 * a readable and writeable memory buffer. It supports reading and writing
 * at the same time by keeping two independent index positions but is not
 * implemented in a thread-safe way, therefore it should be used by multiple
 * threads concurrently.
 */
public interface MemoryBuffer
        extends ReadableMemoryBuffer, WritableMemoryBuffer {

    /**
     * Returns the current capacity of this buffer.
     *
     * @return the current capacity
     */
    int capacity();

    /**
     * Returns the maximum capacity of this buffer.
     *
     * @return the maximum capacity
     */
    int maxCapacity();

    /**
     * Removes all content from this buffer and resets the index positions for
     * reader and writer index to 0.
     */
    void clear();

    /**
     * Returns a duplicate of this buffer, sharing the same underlying content
     * area. It can be used as an independent view of the content but changes
     * to the original or the duplicate will also be reflected to the other buffer.
     *
     * @return a new view of the same content with independent indexes
     */
    MemoryBuffer duplicate();

    /**
     * Acquires a lock on the buffer to prevent garbage collection or eager freeing
     * of native memory areas.
     */
    void lock();

    /**
     * Releases a previously acquired lock.
     */
    void release();

    /**
     * Returns <tt>true</tt> if the buffer is already released, otherwise it returns
     * <tt>false</tt>.
     *
     * @return true if buffer is released, otherwise false
     */
    boolean isReleased();

    /**
     * Returns <tt>true</tt> if all locks are either freed or no lock was hold on the buffer,
     * otherwise it returns <tt>false</tt>.
     *
     * @return true if the buffer is releasable, otherwise false
     */
    boolean isReleasable();

}

File: javascript/src/tengi-core/com/noctarius/tengi/spi/buffer/MemoryBuffer.ts
module com.noctarius.spi.buffer {
    export interface MemoryBuffer {
        capacity(): number;
        maxCapacity(): number;
        clear(): void;
        duplicate(): MemoryBuffer;
        lock(): void;
        release(): void;
        isReleased(): boolean;
        isReleasable(): boolean;

        readable(): boolean;
        readableBytes(): number;
        readBytes(bytes:Uint8Array, offset:number, length:number): void;
        readBuffer(buffer:MemoryBuffer, offset:number, length:number): void;
        readByte(): number;
        readerIndex: number;

        writable(): boolean;
        writableBytes(): number;
        writeBytes(bytes:Uint8Array, offset:number, length:number): void;
        writeBuffer(buffer:MemoryBuffer, offset:number, length:number): void;
        writeByte(value:number): void;
        writerIndex: number;
    }

    export function create(param:any):MemoryBuffer {
        if (param instanceof Uint8Array) {
            return new MemoryBufferImpl(<Uint8Array> param);
        } else if (param instanceof ArrayBuffer) {
            return new MemoryBufferImpl(new Uint8Array(param));
        } else if (typeof(param) === "number") {
            return new MemoryBufferImpl(new Uint8Array(<number> param));
        }
        throw Error("Illegal param type, allowed [Uint8Array, ArrayBuffer, number]: " + typeof(param));
    }

    class MemoryBufferImpl implements buffer.MemoryBuffer {
        private static MAX_LENGTH:number = 65353 * 1000;

        private _refCounter:number = 0;
        private _readerIndex:number;
        private _writerIndex:number;
        private _buffer:Uint8Array;

        constructor(buffer:Uint8Array) {
            this._buffer = buffer;
        }

        get readerIndex():number {
            return this._readerIndex;
        }

        set readerIndex(readerIndex:number) {
            this._readerIndex = readerIndex;
        }

        get writerIndex():number {
            return this._writerIndex;
        }

        set writerIndex(writerIndex:number) {
            this._writerIndex = writerIndex;
        }

        capacity():number {
            return this._buffer.length;
        }

        maxCapacity():number {
            return MemoryBufferImpl.MAX_LENGTH;
        }

        clear():void {
            if (this._buffer != null) this._buffer.length = 1;
        }

        duplicate():buffer.MemoryBuffer {
            return new MemoryBufferImpl(this._buffer);
        }

        lock():void {
            this._refCounter++;
        }

        release():void {
            var refCounter = this._refCounter;
            if (refCounter == 0) {
                return;
            }
            this._refCounter--;
            if (refCounter == 1) {
                this.clear();
            }
        }

        isReleased():boolean {
            return this._refCounter <= 0;
        }

        isReleasable():boolean {
            return this._refCounter == 1;
        }

        readable():boolean {
            return this._buffer.length < this.readerIndex;
        }

        readableBytes():number {
            return this.writerIndex - this.readerIndex;
        }

        readBytes(bytes:Uint8Array, offset:number, length:number):void {
            for (var i:number = 0; i < length; i++) {
                bytes[offset + i] = this.readByte();
            }
        }

        readBuffer(buffer:buffer.MemoryBuffer, offset:number, length:number):void {
            var mb = <MemoryBufferImpl> buffer;
            for (var i:number; i < length; i++) {
                mb.writeByte(this._buffer[offset + i]);
            }
        }

        readByte():number {
            return this._buffer[this._readerIndex++];
        }

        writable():boolean {
            return this.writerIndex < this._buffer.length;
        }

        writableBytes():number {
            return this._buffer.length - this.writerIndex;
        }

        writeBytes(bytes:Uint8Array, offset:number, length:number):void {
            for (var i:number; i < length; i++) {
                this.writeByte(bytes[offset + i]);
            }
        }

        writeBuffer(buffer:buffer.MemoryBuffer, offset:number, length:number):void {
            var mb = <MemoryBufferImpl> buffer;
            for (var i:number; i < length; i++) {
                this.writeByte(mb._buffer[offset + i]);
            }
        }

        writeByte(value:number):void {
            this._buffer[this._writerIndex++] = value;
        }
    }
}